crates/pessimistic-proof/src/certificate.rs

```diff
@@ -1,41 +1,149 @@
     pub bridge_exits: Vec<BridgeExit>,
-    /// The set of imported bridge exits for which the origin network is the target.
-    pub imported_bridge_exits: Vec<BridgeExit>,
+    pub imported_bridge_exits: Vec<ImportedBridgeExit>,
+    pub signature: Signature,
 }
 
 impl Certificate {
-    /// Creates a new [`Certificate`].
-    pub fn new(
-        origin_network: NetworkId,
-        prev_local_exit_root: Digest,
-        bridge_exits: Vec<BridgeExit>,
-    ) -> Self {
+    pub fn new_for_test(network_id: NetworkId, height: Height) -> Self {
         Self {
-            origin_network,
-            prev_local_exit_root,
-            bridge_exits,
-            imported_bridge_exits: Default::default(),
+            network_id,
+            height,
+            prev_local_exit_root: [0; 32],
+            new_local_exit_root: [0; 32],
+            bridge_exits: Vec::new(),
+            imported_bridge_exits: Vec::new(),
+            signature: Signature {
+                r: U256::ZERO,
+                s: U256::ZERO,
+                odd_y_parity: false,
+            },
         }
     }
+
+    pub fn hash(&self) -> Digest {
+        keccak256_combine([
+            self.network_id.to_be_bytes().as_slice(),
+            self.height.to_be_bytes().as_slice(),
+            self.prev_local_exit_root.as_slice(),
+            self.new_local_exit_root.as_slice(),
+        ])
+    }
+
+    pub fn into_pessimistic_proof_input(
+        &self,
+        state: &LocalNetworkStateData,
+        signer: Address,
+    ) -> Result<MultiBatchHeader<Keccak256Hasher>, Error> {
+        let prev_balance_root = state.balance_tree.root;
+        let prev_nullifier_root = state.nullifier_set.root;
+
+        let balances_proofs: BTreeMap<TokenInfo, (U256, LocalBalancePath<Keccak256Hasher>)> = {
+            let mut res = BTreeMap::new();
+
+            let mutated_tokens = {
+                let imported_tokens = self
+                    .imported_bridge_exits
+                    .iter()
+                    .map(|exit| exit.bridge_exit.token_info);
+                let exported_tokens = self.bridge_exits.iter().map(|exit| exit.token_info);
+                imported_tokens.chain(exported_tokens)
+            };
+
+            mutated_tokens.for_each(|token| {
+                let initial_balance =
+                    U256::from_be_bytes(state.balance_tree.get(token).unwrap_or_default());
+                let path = if initial_balance.is_zero() {
+                    // TODO: dont clone once get_inclusion_proof_zero doesnt mutate anymore
+                    state
+                        .balance_tree
+                        .clone()
+                        .get_inclusion_proof_zero(token)
+                        .unwrap()
+                } else {
+                    state.balance_tree.get_inclusion_proof(token).unwrap()
+                };
+                res.insert(token, (initial_balance, path));
+            });
+            res
+        };
```

Comment:
Minor: It seems to me it should be possible to use `mutated_tokens.map(|token| ...).collect()` here in place of `for_each`/`insert`.